beluga/include/beluga/sensor/ndt_sensor_model.hpp

```diff
@@ -0,0 +1,244 @@
+      reinterpret_cast<const Scalar*>(points.data()), 2, static_cast<int64_t>(points.size()));
+  const Eigen::Matrix<Scalar, NDim, 1> mean = points_view.rowwise().mean();
+  const Eigen::Matrix<Scalar, NDim, Eigen::Dynamic> centered = points_view.colwise() - mean;
+  // Use sample covariance.
+  Eigen::Matrix<Scalar, NDim, Eigen::Dynamic> cov =
+      (centered * centered.transpose()) / static_cast<double>(points_view.cols() - 1);
+  cov(0, 0) = std::max(cov(0, 0), kMinVariance);
+  cov(1, 1) = std::max(cov(1, 1), kMinVariance);
+  if constexpr (NDim == 3) {
+    cov(2, 2) = std::max(cov(2, 2), kMinVariance);
+  }
+  return NDTCell<NDim, Scalar>{mean, cov};
+}
+
+/// Given a number of N dimensional points and a resolution, constructs a vector of NDT cells, by clusterizing the
+/// points using 'resolution' and fitting a normal distribution to each of the resulting clusters if they contain a
+/// minimum number of points in them.
+template <int NDim, typename Scalar = double>
+inline std::vector<NDTCell<NDim, Scalar>> to_cells(
+    const std::vector<Eigen::Matrix<Scalar, NDim, 1>>& points,
+    const double resolution) {
+  static constexpr int kMinPointsPerCell = 5;
+
+  Eigen::Map<const Eigen::Matrix<Scalar, NDim, Eigen::Dynamic>> points_view(
+      reinterpret_cast<const Scalar*>(points.data()), NDim, static_cast<int64_t>(points.size()));
+
+  std::vector<NDTCell<NDim, Scalar>> ret;
+  ret.reserve(static_cast<size_t>(points_view.cols()) / kMinPointsPerCell);
+
+  std::unordered_map<Eigen::Matrix<int, NDim, 1>, std::vector<Eigen::Matrix<Scalar, NDim, 1>>, CellHasher<NDim>>
+      cell_grid;
+  for (const Eigen::Matrix<Scalar, NDim, 1>& col : points) {
+    cell_grid[(col / resolution).template cast<int>()].emplace_back(col);
+  }
+
+  for (const auto& [cell, points_in_cell] : cell_grid) {
+    if (points_in_cell.size() < kMinPointsPerCell) {
+      continue;
+    }
+    ret.push_back(fit_points<NDim, Scalar>(points_in_cell));
+  }
+
+  return ret;
+}
+
+}  // namespace detail
+
+/// Parameters used to construct a NDTSensorModel instance.
+struct NDTModelParam {
+  /// Likelihood for measurements that lie inside cells that are not present in the map.
+  double minimum_likelihood = 0;
+  /// Scaling parameter d1 in literature, used for scaling 2D likelihoods.
+  double d1 = 1.0;
+  /// Scaling parameter d2 in literature, used for scaling 2D likelihoods.
+  double d2 = 1.0;
+};
+
+/// NDT sensor model for range finders.
+/**
+ * This class satisfies \ref SensorModelPage.
+ *
+ * \tparam SparseGridT Type representing a sparse NDT grid, as a specialization of 'beluga::SparseValueGrid'.
+ */
+template <typename SparseGridT>
+class NDTSensorModel {
+ public:
+  static_assert(std::is_same_v<SparseValueGrid<typename SparseGridT::map_type>, SparseGridT>);
+  /// NDT Cell type.
+  using ndt_cell_type = typename SparseGridT::mapped_type;
+  static_assert(
+      ndt_cell_type::num_dim == 2 || ndt_cell_type::num_dim == 3,
```

Comment:
@serraramiro1 should we constrain this to 2D only until the rest of the implementation catches up?